server/app/views/questiontypes/ApplicantQuestionRendererImpl.java

```diff
@@ -78,23 +83,63 @@ public final DivTag render(ApplicantQuestionRendererParams params) {
         validationErrors.getOrDefault(question.getContextualizedPath(), ImmutableSet.of());
     if (!questionErrors.isEmpty()) {
       // Question error text
-      questionTextDiv.with(
+      questionSecondaryTextDiv.with(
           BaseHtmlView.fieldErrors(
               messages, questionErrors, ReferenceClasses.APPLICANT_QUESTION_ERRORS));
     }
 
     if (question.isRequiredButWasSkippedInCurrentProgram()) {
       String requiredQuestionMessage = messages.at(MessageKey.VALIDATION_REQUIRED.getKeyName());
-      questionTextDiv.with(
+      questionSecondaryTextDiv.with(
           div()
               .withClasses(Styles.P_1, Styles.TEXT_RED_600)
               .withText("*" + requiredQuestionMessage));
     }
 
+    // Composite fields should be rendered with fieldset and legend for screen reader users.
+    ContainerTag questionPrimaryTextTag;
+    ImmutableList<DomContent> questionTextDoms =
+        TextFormatter.createLinksAndEscapeText(
+            question.getQuestionText(), TextFormatter.UrlOpenAction.NewTab);
+    switch (inputFieldType) {
+      case COMPOSITE:
+        questionPrimaryTextTag = legend().with(questionTextDoms);
+        break;
+      case SINGLE:
+        questionPrimaryTextTag = div().with(questionTextDoms);
+        break;
+      default:
+        throw new IllegalArgumentException(
+            String.format("Unhandled input field type: %s", inputFieldType));
+    }
+    questionPrimaryTextTag.withClasses(
+        ReferenceClasses.APPLICANT_QUESTION_TEXT, ApplicantStyles.QUESTION_TEXT);
+
+    ContainerTag questionTag;
+    switch (inputFieldType) {
+      case COMPOSITE:
+        // Legend must be a direct child of fieldset for screen readers to work properly.
+        questionTag =
+            fieldset()
+                .with(questionPrimaryTextTag)
+                .with(questionSecondaryTextDiv)
+                .with(renderTag(params, validationErrors));
+        break;
+      case SINGLE:
+        questionTag =
+            div()
+                .with(questionPrimaryTextTag)
+                .with(questionSecondaryTextDiv)
+                .with(renderTag(params, validationErrors));
+        break;
+      default:
+        throw new IllegalArgumentException(
+            String.format("Unhandled input field type: %s", inputFieldType));
+    }
```

Comment:
suggestion (non-blocking): why not have a single switch and set both variables in it?

Reply:
Yeah I was trying to maximize code sharing so it would be easier to keep in sync, but I think all I'm doing is saving one line (`.withClasses(ReferenceClasses.APPLICANT_QUESTION_TEXT, ApplicantStyles.QUESTION_TEXT)`). Sounds good to remove the extra switch block.